### PR TITLE
Use sync.Pool for gzipWriter

### PR DESCRIPTION
**Pull request by glefloch**
This PR aims to close issue #366 by adding a sync.Pool for `gzipWriter`. 
As mentionned by @dswarbrick, the `NYTime gzipHandler` did not improve performance compared to a `sync.Pool`. 

Signed-off-by: glefloch <glfloch@gmail.com>

### CODE REVIEW

**Comment by dswarbrick** [comment]
Is there any performance (or GC) difference having the gzipHandler as an anonymous function, vs. declared normally?

File: prometheus/promhttp/http.go [hunk 1]
```diff
@@ -133,10 +124,9 @@ func HandlerFor(reg prometheus.Gatherer, opts HandlerOpts) http.Handler {
 		}
 
 		contentType := expfmt.Negotiate(req.Header)
-		buf := getBuf()
-		defer giveBuf(buf)
-		writer, encoding := decorateWriter(req, buf, opts.DisableCompression)
-		enc := expfmt.NewEncoder(writer, contentType)
+		buf := &bytes.Buffer{}
```
Comment (beorn7) [inline prometheus/promhttp/http.go]: Why did you remove the sync.Pool for buf? This will create a lot of allocations in high-scrape-rate scenarios.

Comment (beorn7) [reply prometheus/promhttp/http.go]: Or to put it another way: For the same reason we want the gzip writer in a sync.Pool, we want this intermediate buffer in a sync.Pool.

File: prometheus/promhttp/http.go [hunk 2]
```diff
@@ -155,29 +145,50 @@ func HandlerFor(reg prometheus.Gatherer, opts HandlerOpts) http.Handler {
...
-			closer.Close()
-		}
+
 		if lastErr != nil && buf.Len() == 0 {
 			http.Error(w, "No metrics encoded, last error:\n\n"+lastErr.Error(), http.StatusInternalServerError)
 			return
 		}
-		header := w.Header()
-		header.Set(contentTypeHeader, string(contentType))
-		header.Set(contentLengthHeader, fmt.Sprint(buf.Len()))
```
Comment (beorn7) [inline prometheus/promhttp/http.go]: This commit removes this header (which SHOULD be used as per https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.13 ). That's probably OK for gzip'd case because we avoid another buffering layer, but it could still be set in the cases where we don't compress. This is complicated here because of the double-wrapping, see comment below.

File: prometheus/promhttp/http.go [hunk 3]
```diff
@@ -155,29 +145,50 @@ func HandlerFor(reg prometheus.Gatherer, opts HandlerOpts) http.Handler {
...
+
+		w.Header().Set(contentTypeHeader, string(contentType))
+
 		if _, err := w.Write(buf.Bytes()); err != nil && opts.ErrorLog != nil {
 			opts.ErrorLog.Println("error while sending encoded metrics:", err)
 		}
 		// TODO(beorn7): Consider streaming serving of metrics.
 	})
 
+	gzipHandler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```
Comment (beorn7) [inline prometheus/promhttp/http.go]: It is helpful to have two `http.HandlerFunc`s here where one just wraps the other? Couldn't this just be inlined?

**Comment by beorn7** [comment]
Thanks for doing this. The idea to gzip the completely rendered response instead of the decorated writer approach so far looks quite neat.

I left a few comments we need to clarify.

File: prometheus/promhttp/http.go (hunk 3 above)
Comment (dswarbrick) [reply prometheus/promhttp/http.go]: Nested handlers is the normal way to implement gzip middleware in Go, especially if you want it to be optional / settable by some flag.

Comment (dswarbrick) [reply prometheus/promhttp/http.go]: I'm not a fan of it being an anonymous function either, as I suspect this will cause a lot of allocations / GC. The NYTimes implementation is worth studying. And bear in mind that the content-length header should be the length of the gzipped output, not the original input.

Comment (beorn7) [reply prometheus/promhttp/http.go]: WRT content length: Yes, I had that in mind. That's why I suggested it's probably OK to not add a content length when gzipping. But we should (or even SHOULD :-) still add the content length when _not_ gzipping. And that's not possible with the approach here where the decision is made within the gzipHandler local function but the uncompressed content length is only available in the h local function. In this specific case, where we don't want to provide a generically usable gzip handler, I'd probably just inline everything into one local function. At least for now. We can still refactor into something nicer once we have something that works.

**Comment by glefloch** [comment]
@beorn7 Thanks for the comment, I will update the code.

**Comment by beorn7** [comment]
Hmm, thinking about the need (or not-need) of content length, the changes proposed in your PR, and the current performance testing I'm doing for kube-state-metrics, I'm wondering if we should perhaps ditch that intermediate buffer and do streaming exposition, at least for the encoding part.

I'll play with it and create a PR based on yours.

**Comment by beorn7** [comment]
See https://github.com/prometheus/client_golang/pull/482
